cadquery/occ_impl/solver.py

```diff
@@ -20,36 +62,262 @@
+            )
+
+    def _getAxis(self, arg: Shape) -> gp_Dir:
+
+        if isinstance(arg, Face):
+            rv = arg.normalAt()
+        elif isinstance(arg, Edge) and arg.geomType() != "CIRCLE":
+            rv = arg.tangentAt()
+        elif isinstance(arg, Edge) and arg.geomType() == "CIRCLE":
+            rv = arg.normal()
+        else:
+            raise ValueError(f"Cannot construct Axis for {arg}")
+
+        return rv.toDir()
+
+    def _getPln(self, arg: Shape) -> gp_Pln:
+
+        if isinstance(arg, Face):
+            rv = gp_Pln(self._getPnt(arg), arg.normalAt().toDir())
+        elif isinstance(arg, (Edge, Wire)):
+            normal = arg.normal()
+            origin = arg.Center()
+            plane = Plane(origin, normal=normal)
+            rv = plane.toPln()
+        else:
+            raise ValueError(f"Can not construct a plane for {arg}.")
+
+        return rv
+
+    def _getPnt(self, arg: Shape) -> gp_Pnt:
+
+        # check for infinite face
+        if isinstance(arg, Face) and any(
+            Precision.IsInfinite_s(x) for x in BRepTools.UVBounds_s(arg.wrapped)
+        ):
+            # fall back to gp_Pln center
+            pln = arg.toPln()
+            center = Vector(pln.Location())
+        else:
+            center = arg.Center()
+
+        return center.toPnt()
+
+    def toPODs(self) -> Tuple[Constraint, ...]:
+        """
+        Convert the constraint to a representation used by the solver.
+
+        NB: Compound constraints are decomposed into simple ones.
+        """
+
+        # apply sublocation
+        args = tuple(
+            arg.located(loc * arg.location())
+            for arg, loc in zip(self.args, self.sublocs)
+        )
+
+        markers: List[Tuple[ConstraintMarker, ...]]
+
+        # convert to marker objects
+        if self.kind == "Axis":
+            markers = [(self._getAxis(args[0]), self._getAxis(args[1]),)]
+
+        elif self.kind == "Point":
+            markers = [(self._getPnt(args[0]), self._getPnt(args[1]))]
+
+        elif self.kind == "FixedPoint":
+            markers = [(self._getPnt(args[0]),)]
+
+        elif self.kind == "Plane":
+            markers = [
+                (self._getAxis(args[0]), self._getAxis(args[1]),),
+                (self._getPnt(args[0]), self._getPnt(args[1])),
+            ]
+
+        elif self.kind == "PointInPlane":
+            markers = [(self._getPnt(args[0]), self._getPln(args[1]))]
+
+        elif self.kind == "Fixed":
+            markers = [(None,)]
+
+        elif self.kind == "FixedPoint":
+            markers = [(self._getPnt(args[0]),)]
+
+        elif self.kind == "FixedAxis":
+            markers = [(self._getAxis(args[0]),)]
+
+        else:
+            raise ValueError(f"Unknown constraint kind {self.kind}")
+
+        # specify kinds of the simple constraint
+        if self.kind in CompoundConstraints:
+            kinds = CompoundConstraints[self.kind]
+            params = (self.param,) * len(kinds)
```

Comment:
Readability does look to be much improved.  I'm still getting familiar with the assy solver.

Regarding `CompoundConstraints` (Plane), here the same `param` value is passed to both the Axis and Point simple constraints.

Sharing the same `param` value across the simple constraints (at least for Plane) didn't make sense to me.  It could be I'm missing something.  I thought I'd explore passing two `param` values to `CompoundConstraints` (3.) or for Plane, fix the Point `param` and allow the Axis `param` angle to vary (2.). 


1. Single `param` value passed to all simple constraints

   This is the current implementation.  Say I specify a Plane constraint with `param=pi/2`.

   This solves for Axis angle of pi/2, and Point **distance** pi/2.  It doesn't make sense to me because the Axis param value is an angle and the Point param is a distance.

   If a single `param` is passed to both Axis, and Point, then any value other than 0, or None (both already treated as 0 for Point param) probably should be disallowed.

   When I try Plane with value other than 0, or None, I get very weird results because the solver optimizes for Point distance set to the angle in radians. 


2. Pass `param` to Axis, fix Point `param` to 0

   For more flexibility of the Plane constraint, perhaps allow for passing angle param value of 0 to pi.  The Point param value would always be fixed to zero. 

   `params = (self.param, 0)`

   Plane could then be used for any combined Axis, Point case unless non-zero Point param was required.


3. Pass two separate `param` values to the individual simple constraints

   Perhaps some backwards compatible way could be designed to pass a tuple of param values to Plane.

   Say a Plane constraint is specified with `param=(pi,100)`.

   This might offer the most flexibility for Plane without requiring to enter two separate Axis and Point constraints.

Reply:
Good point, I think I'd go for 3.

Reply:
After some thinking I implemented 2 actually.